File: di_python/src/mas/ServiceImplMAS.py
"""
Dependency Injection project in Python

This Python project demonstrates dependency injection using constructor injection.
It includes an API with two interface classes: DAO and Service.
The DAO interface simulates data access, while the Service interface simulates data computation.
Two versions (mas and hr) implement constructor injection from DAO into Service layer.

ServiceImplMAS Class Documentation

This Python class implements the Service interface to provide computation functionality for version MAS
of the project.

Author: Dac Linh Giao NGUYEN
Date: 2024

Usage:
    - Instantiate ServiceImplMAS with a DAO instance to perform computations using data from the MAS data source.
    - Call the compute method to perform computation.
"""

from src.api.Service import Service


class ServiceImplMAS(Service):
    """
    Concrete implementation of the Service interface for version 1 of the project.
    This implementation performs computation based on data retrieved from a DAO instance.
    """

    def __init__(self, dao):
        """
        Constructor for ServiceImplMAS.
        Inject the specified DAO in ServiceImplMAS.

        Parameters:
            dao (DAO): An instance of a data access object.
        """
        self.dao = dao

    def compute(self):
        """
        Perform computation based on data retrieved from the DAO instance.

        Returns:
            float: The result of the computation.
        """
        print("Service computes 70% of MAS")
        data = self.dao.get_data()
        return data * 0.70

File: di_kotlin/src/main/kotlin/mas/ServiceImplMAS.kt
/*
 * Dependency injection project.
 * Dac Linh Giao NGUYEN
 * License MIT
 *
 * Dependency injection project in Kotlin.
 * ServiceImplMAS.kt
 */
package com.kmmania.mas

import com.kmmania.api.DAO
import com.kmmania.api.Service

/**
 * Implementation of Service interface for MAS (Maximal Aerobic Speed) data.
 *
 * This service computes a percentage of the Maximal Aerobic Speed.
 *
 * @param dao the DAO implementation providing access to MAS data.
 * @author Dac Linh Giao NGUYEN
 * @since 0.1
 * @see Service
 */
class ServiceImplMAS(private val dao: DAO) : Service {
    /**
     * Computes 70% of the Maximal Aerobic Speed.
     *
     * @return 70% of the MAS as a Double.
     */
    override fun compute(): Double {
        println("Service computes 70% of MAS")
        val data = dao.getData()
        return data * 0.70
    }
}